packages/react/src/JsonFormsContext.tsx

```diff
@@ -144,6 +189,89 @@ export const JsonFormsStateProvider = ({
       })
     )
   );
+
+  useEffect(() => {
+    coreDispatch(
+      Actions.updateCore(data, schema, uischema, {
+        ajv,
+        validationMode,
+        additionalErrors,
+      })
+    );
+  }, [data, schema, uischema, ajv, validationMode, additionalErrors]);
+
+  /**
+   * A common pattern for users of JSON Forms is to feed back the data which is emitted by
+   * JSON Forms to JSON Forms ('controlled style').
+   *
+   * Every time this happens, we dispatch the 'updateCore' action which will be a no-op when
+   * the data handed over is the one which was just recently emitted. This allows us to skip
+   * rerendering for all normal cases of use.
+   *
+   * However there can be extreme use cases, for example when using Chrome Auto-fill for forms,
+   * which can cause JSON Forms to emit multiple change events before the parent component is
+   * rerendered. Therefore not the very recent data, but the previous data is fed back to
+   * JSON Forms first. JSON Forms recognizes that this is not the very recent data and will
+   * validate, rerender and emit a change event again. This can then lead to data loss or even
+   * an endless rerender loop, depending on the emitted events chain.
+   *
+   * To handle these edge cases in which many change events are sent in an extremely short amount
+   * of time we debounce them over a short amount of time. 10ms was chosen as this worked well
+   * even on low-end mobile device settings in the Chrome simulator.
+   */
+  const debouncedEmit = useCallback(
+    debounce((...args: any[]) => onChangeRef.current?.(...args), 10),
+    []
+  );
+
+  return (
+    <JsonFormsDataProvider
+      initState={initState}
+      onChangeHandler={debouncedEmit}
+      core={core}
+      coreDispatch={coreDispatch}
+    >
+      {children}
+    </JsonFormsDataProvider>
+  );
+};
+
+const JsonFormsControlledStateProvider = ({
+  children,
+  initState,
+  onChangeRef,
+}: JsonFormsStateProviderProps & { onChangeRef: any }) => {
+  const { data, schema, uischema, ajv, validationMode, additionalErrors } =
+    initState.core;
+
+  const onChangeHandler = useCallback(
+    (...args: any[]) => onChangeRef.current?.(...args),
+    []
+  );
+
+  const [core, coreDispatch] = useReducer(coreReducer, undefined, () =>
+    coreReducer(
+      initState.core,
+      Actions.init(data, schema, uischema, {
+        ajv,
+        validationMode,
+        additionalErrors,
+      })
+    )
+  );
+
+  const actualDispatch = (action: CoreActions) => {
+    if (action.type === 'jsonforms/UPDATE') {
+      const updatedState = coreReducer(core, action);
+      return onChangeHandler({
+        data: updatedState.data,
+        errors: core.errors,
+      });
```

Comment:
Ah interesting, this is not only emitting the data but actually is some kind of `middleware` support, only for the `jsonforms/UPDATE` case. However the user has no chance to actually return something sensible in the `onChangeHandler` here. To not update the internal state we need to return the `core` here, but the user does not have access to it.

Hmm thinking about it, we could actually just introduce a generic middleware support and then this use case of controlled vs uncontrolled mode would also be covered.

This could look like this:
```javascript
const middleware = props.middleware ?? (core, action, coreDispatch) => coreDispatch(core, action);

const actualDispatch = (action: CoreActions) => {
  return middleware(core, action, coreDispatch);
}
```

The middleware for this case could then look like this:

```javascript
const controlledModeMiddleware = (state, action, defaultDispatch) => {
  if(action.type === 'jsonforms/UPDATE'){
     const updatedState = defaultDispatch(state, action);
     myChangeHandler(data: updatedState.data, errors: updatedState.errors);
     return state;
   } else {
     return defaultDispatch(state, action);
   }
}
```

The new config could then be used to disable the default (debounced) emitter. Although not even that is needed as the `onChange` handler does not need to be handed over in that case.

Reply:
Hi @sdirix 👋.

As I mentioned on a couple of my PR review comments, there's a lot of repetition between the different state providers, and one piece of mode related logic is not currently encapsulated within the relevant state provider, where it would ideally reside.

I just performed some diffs between the mixed, controlled and uncontrolled state providers to see what the lie of the land was, and this is what I determined:

1. Mixed debounces `onChange`.
2. Uncontrolled passes `core.data` to `coreDispatch(Actions.updateCore)`, whereas Mixed and Controlled pass `data`.
3. Controlled returns a wrapped dispatch function rather than the underlying dispatch function.
4. Controlled 'effectively' passes a dummy `onChange`:
   1. This is currently implemented within `JsonFormsDataProvider`, which only invokes `onChange` if `mode !== 'controlled'`, but it makes more sense that all mode logic be centralised within the data provider.

This means that there are three points of variance that any middleware (or data-provider) would have to account for:

1. Whether the internally or externally sourced data is passed to the reducer.
2. Whether `onChange` is wrapped.
3. Whether `dispatch` is wrapped.

As I understand it, the middleware idea would only help with the last of these three, whereas the data provider can handle all three.

Although the state providers currently have a lot of repetition between them, I've taken a look and this can be dramatically reduced by creating a couple of shared hooks (e.g. `useCoreDispatcher` and `useCoreUpdater`) to encapsulate the common blocks of code. If we also take the opportunity to move the state providers into their own files, then that would clean up things even further.

Let me know if you agree or not, and I (or @dholmes2409) can take a look at getting the PR updated as described.

Reply:
Hi @dchambers, thanks for the summary! I thought we could efficiently tackle the "mixed" vs "controlled" case by only implementing the suggested middleware approach without the need to introduce multiple state providers. However I might have also overlooked some edge case / use case. If only "mixed" and "controlled" is offered, would this be sufficient for your use cases, or would you specifically require all three modes?

I will take a more detailed look at your analysis and my suggestion early next week and then come back to you with the go ahead for the middleware or the different state providers. Overall planning wise I'm pretty confident that we will get this into a November release.

Reply:
> Hi @dchambers, thanks for the summary! I thought we could efficiently tackle the "mixed" vs "controlled" case by only implementing the suggested middleware approach without the need to introduce multiple state providers. However I might have also overlooked some edge case / use case. If only "mixed" and "controlled" is offered, would this be sufficient for your use cases, or would you specifically require all three modes?

Hi @sdirix 👋, I hope you had a good weekend. "controlled" is the only mode we'd really want to use, so this would definitely be sufficient for us if it can be made to work. That said, I'm not currently clear as to how a middleware could replace the `onChange` that ends up being passed to `JsonFormsDataProvider`, if you're able to provide some clarity here?

Reply:
Hi @dchambers!

I quickly implemented my suggestion. However I did not yet test nor compile it. You can find it in my fork, see [here](https://github.com/sdirix/jsonforms/tree/middleware-support).

Please have a look. Using a middleware you no longer need to hand over a "onChange" as you get the latest updates anyway. Therefore the `debounce` is also avoided.

Reply:
Hi @sdirix 👋. Thanks for this, and the code looks really clean 👌.

> Using a middleware you no longer need to hand over a "onChange" as you get the latest updates anyway. Therefore the debounce is also avoided.

Sorry, I'm struggling to understand this bit. Your branch has an implementation of `defaultMiddleware`, but no implementation of `controlledMiddleware`, yet this would probably help to clarify things. Here's a stab at what I think this might look like:

```ts
const controlledMiddleware: Middleware = (state, action, defaultDispatch) => {
  if (action.type === 'jsonforms/UPDATE') {
    const [updatedState, errors] = reducer(state, action);
    return onChange({
      data: updatedState,
      errors,
    });
  } else {
    return defaultDispatch(state, action);
  }
};
```

but this begs the question: "Where do `reducer` and `onChange` come from?".

Unless I'm misunderstanding something, I'd therefore expect the middleware interface to get access to more of the same constants that the data provider component had access to. Consider the following pseudcode, for example:

```ts
export type MiddlewareDispatcher = (
  state: JsonFormsCore,
  action: CoreActions
) => JsonFormsCore;

export type MiddlewareOptions = {
  reducer: (
    state: JsonFormsCore,
    action: CoreActions
  ) => [JsonFormsCore, JsonFormsErrors];
  onChange: (update: { data: JsonFormsCore; errors: JsonFormsErrors }) => void;
  defaultDispatcher: MiddlewareDispatcher;
};

export type Middleware = (
  state: JsonFormsCore,
  action: CoreActions,
  options: MiddlewareOptions
) => JsonFormsCore;

export const defaultMiddleware: Middleware = (state, action, options) =>
  options.defaultDispatcher(state, action);

export const controlledMiddleware: Middleware = (state, action, options) => {
  if (action.type === 'jsonforms/UPDATE') {
    const [updatedState, errors] = options.reducer(state, action);
    options.onChange({
      data: updatedState,
      errors,
    });

    return updatedState;
  } else {
    return options.defaultDispatcher(state, action);
  }
};
```

Thoughts?

Reply:
Hi @dchambers,

I didn't add the `controlledMiddleware` to the repository as I expect it to be implemented in JSON Forms consumers. If more people have this use case we could maybe offer a factory util to create such a middleware conveniently.

Based on your example I think this can look like this:

```typescript
/**
 * Controlled variant of JSON Forms.
 * Props are only read in first render pass. Updates to them are ignored.
 */
const ControlledJsonForms = ({ schema, initialData }) => {
  const [formState, setFormState] = useState<{
    data: any;
    errors: ErrorObject[];
  }>({ data: initialData, errors: [] });

  const validator = useMemo(() => createAjv().compile(schema), []);

  const controlledMiddleware: Middleware = (state, action, reducer) => {
    if (action.type === 'jsonforms/UPDATE') {
      const newState = reducer(state, action);
      validator(newState.data);
      setFormState({ data: newState.data, errors: validator.errors || [] });
      return state; // this will lead to JSON Forms ignoring the change within itself
    } else {
      return reducer(state, action);
    }
  };

  return (
    <JsonForms
      schema={schema}
      data={formState.data}
      additionalErrors={formState.errors}
      validationMode='NoValidation'
      middleware={controlledMiddleware}
      renderers={materialRenderers}
      cells={materialCells}
    />
  );
};
```
The `controlledMiddeware` could be placed in a `useCallback` but it does not have to be.

Overall this looks like I expected, except for the validation part. I realized that if the validation shall also be controlled then it either must be performed outside, like I did now, or it would require more code within the `controlledMiddleware` delving deeper into the internals. As there is no downside in doing the validation on our own I preferred that approach.

What do you think, does this cover your use case?

Reply:
@dchambers Does this work for you? If yes, then I would like to get it in ideally before the next release.

Reply:
Hey @sdirix, I had play around with this today and it seems to cover our use case. We would need to tweak our usage as we pass a custom onChange handler but we could easily move the logic into the controlledMiddleware.
Should I close this PR and we go with the approach you committed to your fork?

Reply:
To get this in faster it would be nice to have some tests. Would you like to contribute them? You could force push this PR and use my commit as basis. In the tests you could add the `controlledMiddleware` and verify that your use case works as expected. Then it's ensured that it will not be broken in the future ;)

If you don't have time then that's also fine, we are definitely eager to get this feature in, so we will eventually also work on this. However as always I can't give any promises regarding a due date as this depends on our work load.

Reply:
Moved this into PR #2220